### PR TITLE
Replaced case sensitive checkbox with a button using an icon

### LINKED ISSUE #645: Unlabeled checkbox for case sensitive search is bad UX
The unlabeled case sensitive search checkbox confused me a lot. At first I ignored it and it took me a while to figure out what it does when suddenly the search was "broken". It is also unclear if checked means case sensitive or insensitive adding to the confusion.

image

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/eclipse/chemclipse/issues/645.

### CODE REVIEW

File: chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java [hunk 1]
```diff
@@ -152,18 +152,20 @@ public void widgetDefaultSelected(SelectionEvent e) {
...
 
-	private Button createCheckBoxCaseSensitive(Composite parent) {
+	private Button createButtonCaseSensitive(Composite parent) {
 
-		Button button = new Button(parent, SWT.CHECK);
+		Button button = new Button(parent, SWT.PUSH);
 		button.setText("");
 		button.setToolTipText("Search Case Sensitive");
-		button.setSelection(PreferenceSupplier.isSearchCaseSensitive());
+		button.setImage(ApplicationImageFactory.getInstance().getImage(IApplicationImage.IMAGE_CASE_SENSITIVE, IApplicationImage.SIZE_16x16));
```
Comment (eselmeister) [inline chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java:169]: @Mailaender When initializing the button image, it should be checked if a case sensitive search shall be performed. Also use the decorator here:

`button.setImage(ApplicationImageFactory.getInstance().getImage(IApplicationImage.IMAGE_CASE_SENSITIVE, IApplicationImage.SIZE_16x16, PreferenceSupplier.isSearchCaseSensitive()));`

**Review by eselmeister — CHANGES_REQUESTED** [review]

**Comment by eselmeister** [comment]
@Mailaender Please have a look at the comment.

File: chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java (hunk 1 above)
Comment (Mailaender) [reply chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java:169]: Added the missing initialization.

File: chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java [hunk 2]
```diff
@@ -152,18 +152,21 @@ public void widgetDefaultSelected(SelectionEvent e) {
...
 
-	private Button createCheckBoxCaseSensitive(Composite parent) {
+	private Button createButtonCaseSensitive(Composite parent) {
 
-		Button button = new Button(parent, SWT.CHECK);
+		Button button = new Button(parent, SWT.PUSH);
 		button.setText("");
 		button.setToolTipText("Search Case Sensitive");
-		button.setSelection(PreferenceSupplier.isSearchCaseSensitive());
+		boolean active = PreferenceSupplier.isSearchCaseSensitive();
```
Comment (eselmeister) [inline chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java]: @Mailaender The boolean variable `active` is only used once here. Probably move the `PreferenceSupplier.isSearchCaseSensitive()` into the set image method:

`button.setImage(ApplicationImageFactory.getInstance().getImage(IApplicationImage.IMAGE_CASE_SENSITIVE, IApplicationImage.SIZE_16x16, PreferenceSupplier.isSearchCaseSensitive()));`

Comment (Mailaender) [reply chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java]: I was worrying about the line length, but you are right.

**Review by eselmeister — APPROVED** [review]

**Comment by eselmeister** [comment]
@Mailaender A final change request:

`private Button caseSensitiveButton;`
Is not used any longer, as the button directly changes the preference. The member variable can be removed and the creation of the button can be simply called via:

`createButtonCaseSensitive(composite); `

I would also recommend to set a tooltip, depending on the status of button, e.g.:
`button.setToolTipText(PreferenceSupplier.isSearchCaseSensitive() ? "Search Case Sensitive" : "Search Normal");`

File: chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java (hunk 1 above)
Comment (Mailaender) [reply chemclipse/plugins/org.eclipse.chemclipse.swt.ui/src/org/eclipse/chemclipse/swt/ui/components/SearchSupportUI.java:169]: I had to revert this due to performance problems.

**Comment by Mailaender** [comment]
Added the tooltip, removed the unused button variable, added boolean caching for performance.

**Comment by eselmeister** [comment]
@Mailaender Thanks. It looks better, but it contains a possibility to get into an inconsistent status. Have a look at the following case:

A) The search UI is created, Case Sensitive = false, the button is not decorated
B) Then the user modifies somewhere else or in the settings the Case Sensitive preference to true
C) Now run a search with the search UI created at A), the button is still not decorated, but the run search is operated with the variable caseSensitive = true. This could lead to confusing results.

Hence, I would recommend to change the code here:

```
@Override
public void widgetSelected(SelectionEvent e) {

	caseSensitive = !caseSensitive;
	PreferenceSupplier.setSearchCaseSensitive(caseSensitive);
	button.setToolTipText(caseSensitive ? "Search Case Sensitive" : "Search Case Insensitive");
	button.setImage(ApplicationImageFactory.getInstance().getImage(IApplicationImage.IMAGE_CASE_SENSITIVE, IApplicationImage.SIZE_16x16, caseSensitive));
	runSearch();
}
```

and here:

```
if(searchListener != null) {
	String searchText = text.getText().trim();
	searchListener.performSearch(searchText, caseSensitive);
}
```
 
Whenever the user changes the search case sensitive flag, the user selection is persisted. But the search UI itself has its own status.

Please let me know if you'd like to discuss this issue.

**Comment by Mailaender** [comment]
Added your changes, but I am still fighting a nightmarish bug here that freezes the whole application. I narrowed it down, but I don't understand why it is a problem. Left a source code comment.

**Comment by eselmeister** [comment]
OK, let me review it.

**Comment by eselmeister** [comment]
That is a tricky one :-).
The plain image needs to be set first. Then, the decorator could come into play.

Have a look at the code to solve the issue:

```
/*******************************************************************************
 * Copyright (c) 2017, 2021 Lablicate GmbH.
 * 
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 * Dr. Philip Wenig - initial API and implementation
 *******************************************************************************/
package org.eclipse.chemclipse.swt.ui.components;

import org.eclipse.chemclipse.model.preferences.PreferenceSupplier;
import org.eclipse.chemclipse.rcp.ui.icons.core.ApplicationImageFactory;
import org.eclipse.chemclipse.rcp.ui.icons.core.IApplicationImage;
import org.eclipse.swt.SWT;
import org.eclipse.swt.events.KeyAdapter;
import org.eclipse.swt.events.KeyEvent;
import org.eclipse.swt.events.SelectionAdapter;
import org.eclipse.swt.events.SelectionEvent;
import org.eclipse.swt.layout.FillLayout;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Button;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Text;

public class SearchSupportUI extends Composite {

	private Text text;
	private Button buttonCaseSensitive;
	//
	private ISearchListener searchListener;
	private boolean caseSensitive = PreferenceSupplier.isSearchCaseSensitive();

	public SearchSupportUI(Composite parent, int style) {

		super(parent, style);
		createControl();
	}

	@Override
	public void setVisible(boolean visible) {

		super.setVisible(visible);
		updateButtonCaseSensitive();
	}

	public void reset() {

		setSearchText("");
	}

	public void setSearchListener(ISearchListener searchListener) {

		this.searchListener = searchListener;
	}

	public void setSearchText(String searchText) {

		text.setText(searchText);
		runSearch();
	}

	/**
	 * Use this e.g. for content proposal listeners.
	 * The search text shall be set via the setSearchText method.
	 * 
	 * @return Text
	 */
	public Text getText() {

		return text;
	}

	public String getSearchText() {

		return text.getText().trim();
	}

	public boolean isSearchCaseSensitive() {

		return caseSensitive;
	}

	private void createControl() {

		setLayout(new FillLayout());
		//
		Composite composite = new Composite(this, SWT.NONE);
		GridLayout gridLayout = new GridLayout(3, false);
		gridLayout.marginLeft = 0;
		gridLayout.marginRight = 0;
		composite.setLayout(gridLayout);
		//
		createButtonSearch(composite);
		text = createTextSearch(composite);
		buttonCaseSensitive = createButtonCaseSensitive(composite);
	}

	private Button createButtonSearch(Composite parent) {

		Button button = new Button(parent, SWT.PUSH);
		button.setText("");
		button.setToolTipText("Run the search");
		button.setImage(ApplicationImageFactory.getInstance().getImage(IApplicationImage.IMAGE_SEARCH, IApplicationImage.SIZE_16x16));
		button.addSelectionListener(new SelectionAdapter() {

			@Override
			public void widgetSelected(SelectionEvent e) {

				runSearch();
			}
		});
		//
		return button;
	}

	private Text createTextSearch(Composite parent) {

		Text text = new Text(parent, SWT.BORDER | SWT.SEARCH | SWT.ICON_CANCEL | SWT.ICON_SEARCH);
		text.setText("");
		text.setToolTipText("Type in the search items.");
		text.setLayoutData(new GridData(GridData.FILL_HORIZONTAL));
		/*
		 * Listen to search key event.
		 */
		text.addKeyListener(new KeyAdapter() {

			@Override
			public void keyReleased(KeyEvent e) {

				if(e.keyCode == SWT.LF || e.keyCode == SWT.CR || e.keyCode == SWT.KEYPAD_CR) {
					runSearch();
				} else if(text.getText().trim().equals("")) {
					/*
					 * Reset when empty.
					 */
					runSearch();
				}
			}
		});
		/*
		 * Click on the icons.
		 */
		text.addSelectionListener(new SelectionAdapter() {

			@Override
			public void widgetDefaultSelected(SelectionEvent e) {

				if(e.detail == SWT.ICON_CANCEL) {
					text.setText("");
					runSearch();
				} else if(e.detail == SWT.ICON_SEARCH) {
					runSearch();
				}
			}
		});
		//
		return text;
	}

	private Button createButtonCaseSensitive(Composite parent) {

		Button button = new Button(parent, SWT.PUSH);
		button.setText("");
		button.setToolTipText("");
		button.setImage(ApplicationImageFactory.getInstance().getImage(IApplicationImage.IMAGE_CASE_SENSITIVE, IApplicationImage.SIZE_16x16));
		button.addSelectionListener(new SelectionAdapter() {

			@Override
			public void widgetSelected(SelectionEvent e) {

				caseSensitive = !caseSensitive;
				PreferenceSupplier.setSearchCaseSensitive(caseSensitive);
				updateButtonCaseSensitive();
				runSearch();
			}
		});
		//
		return button;
	}

	private void updateButtonCaseSensitive() {

		buttonCaseSensitive.setToolTipText(caseSensitive ? "Search Case Sensitive" : "Search Case Insensitive");
		buttonCaseSensitive.setImage(ApplicationImageFactory.getInstance().getImage(IApplicationImage.IMAGE_CASE_SENSITIVE, IApplicationImage.SIZE_16x16, caseSensitive));
	}

	private void runSearch() {

		if(searchListener != null) {
			String searchText = text.getText().trim();
			searchListener.performSearch(searchText, caseSensitive);
		}
	}
}
```

**Comment by Mailaender** [comment]
Interesting. https://github.com/eclipse/chemclipse/blob/984feb6ef50ce1b06e6ab35005575c8353844031/chemclipse/plugins/org.eclipse.chemclipse.ux.extension.xxd.ui/src/org/eclipse/chemclipse/ux/extension/xxd/ui/swt/ExtendedWellChartUI.java#L160 is somehow not effected by this bug.

**Comment by eselmeister** [comment]
The WellChart example is interesting. It could be, that barchart has been used before, hence it's in the cache already. Let's do a further inspection on this.

The SearchSupportUI code looks fine now.

**Pull request merged**